Add sleep scalar function

krleonid opened the pull request:
- Implement sleep() function that takes milliseconds as BIGINT parameter
- Function returns NULL (empty output) after sleeping
- Add comprehensive tests covering various use cases
- Register function in core_functions extension

krleonid committed:
Merge pull request #2 from krleonid/array_intersect

Add list_intersect function implementation

krleonid committed:
Merge branch 'duckdb:main' into main

krleonid committed:
Add sleep scalar function with constant folding optimization

- Implement sleep() function that takes milliseconds as BIGINT parameter
- Function returns NULL (empty output) after sleeping
- Extract sleep function to separate file (sleep.cpp) with binding method
- Add bind_expression callback to execute sleep once at bind time for constant arguments
- Add comprehensive tests covering various use cases
- Register function in core_functions extension

Mytherin commented on extension/core_functions/scalar/debug/sleep.cpp:
```diff
@@ -0,0 +1,72 @@
+
+static unique_ptr<Expression> BindSleepFunctionExpression(FunctionBindExpressionInput &input) {
+	if (input.children[0]->IsFoldable()) {
+		Value value = ExpressionExecutor::EvaluateScalar(input.context, *input.children[0]);
```
I don't think the constant optimization is necessary. In Postgres, for example, the sleep function just runs once for every invocation:

```sql
\timing on
select pg_sleep(1) from generate_series(0,9);
 pg_sleep 
----------
 
 
 
 
 
 
 
 
 
 
(10 rows)

Time: 10011.742 ms (00:10.012)
```

If you want to only sleep once it's easy to only invoke the function once (`select sleep(...)`).

Mytherin commented on extension/core_functions/scalar/debug/sleep.cpp:
```diff
@@ -0,0 +1,72 @@
+
+	auto milliseconds = UnifiedVectorFormat::GetData<int64_t>(vdata);
+	
+	for (idx_t i = 0; i < input.size(); i++) {
```
Could we use `GenericExecutor::ExecuteUnary` here instead of manually looping over the vector?

krleonid replied:
@Mytherin 
The GenericExecutor::ExecuteUnary function contains an optimization for constant vectors here:
https://github.com/duckdb/duckdb/blob/8a28852967fd2907f73ed929cc4828a321deaf9d/src/include/duckdb/common/vector_operations/generic_executor.hpp#L255

So, when I convert the loop over the vector to use it, my code executes only once, regardless of the vector’s size.

Mytherin replied:
We could call `input.Flatten()` before to turn the input into a flat vector

Mytherin reviewed (commented):
Thanks for the PR! Looks good - some comments below. Could we also add an alias to `pg_sleep` in `src/catalog/default/default_functions.cpp` (that would then call e.g. `sleep_ms(x / 1000.0)` so it would take the argument in seconds for Postgres compatibility.

krleonid committed:
Apply clang-tidy fix: remove trailing whitespace

krleonid committed:
Apply code formatting fixes

krleonid committed:
Fix test file group tag: change [function] to [generic]

krleonid committed:
Add sleep_ms function and pg_sleep alias

- Add sleep_ms scalar function that sleeps for specified milliseconds
- Add pg_sleep alias in pg_catalog for Postgres compatibility (takes seconds)
- Refactor to use manual loop instead of GenericExecutor for SQLNULL return type
- Add comprehensive tests for both functions

krleonid committed:
Refactor sleep_ms to use GenericExecutor::ExecuteUnary

- Replace manual loop with GenericExecutor::ExecuteUnary for cleaner code
- Add NullResultType struct to handle SQLNULL return type
- Remove input.Flatten() call as GenericExecutor handles vector types internally

krleonid committed:
Merge branch 'main' of github.com:krleonid/duckdb into sleep_function

maiadegraaf commented on extension/core_functions/scalar/debug/sleep.cpp:
```diff
@@ -0,0 +1,42 @@
+
+namespace duckdb {
+
+namespace {
```
Nit: This doesn't need to be in its own namespace.

maiadegraaf reviewed (commented):
Thanks for implementing the changes! Looks good now, just a few nits and some testing ideas.

krleonid committed:
Refactor sleep_ms function and add comprehensive tests

- Remove redundant name parameter from ScalarFunction constructor (name is already in functions.json)
- Remove unnecessary anonymous namespace wrapper
- Add table-based sleep tests
- Add JSON profiling tests to validate actual timing measurements
- Clean up temporary profiling tables in tests

krleonid committed:
Add upper bound validation to sleep profiling tests

- Add upper bound check (cpu_time <= 0.1) to first profiling test
- Fix SQL syntax: use AND instead of && in second profiling test
- Both tests now validate timing is within reasonable bounds (0.05-0.1s)

maiadegraaf reviewed (approved):
Thanks for the changes! LGTM :)

krleonid commented on test/sql/function/generic/test_sleep.test:
```diff
@@ -0,0 +1,201 @@
+query T
+SELECT cpu_time >= 0.05 FROM sleep_profiling WHERE contains(extra_info, 'sleep_ms') OR contains(query_name, 'sleep_ms') LIMIT 1
+----
+true
```
Thank you for the idea.

krleonid committed:
Add sleep_ms to extension_entries.hpp

lnkuiper commented:
@carlopi Does `this_thread_sleep_for` not work in WASM?

> Also I am not sure what's the behaviour when multiple chunks are executed in parallel (does it sleep also in parallel), but there I am fine with whatever.

Yes, it would sleep in parallel, once per value. I don't think there's a way around this, unless we make this a table function, but I think this is more in line with what people expect (also from PostgreSQL).

carlopi commented:
No, we have stuff like https://github.com/duckdb/duckdb/blob/main/src/main/http/http_util.cpp#L406 where we guard both call and inclusion of `<chrono>`.
I can look at adding this to CI.
